lib/HLSL/HLLowerUDT.cpp

```diff
@@ -274,6 +274,7 @@ void hlsl::ReplaceUsesForLoweredUDT(Value *V, Value *NewV) {
       SmallVector<Value*, 4> idxList(GEP->idx_begin(), GEP->idx_end());
       Value *NewGEP = Builder.CreateGEP(NewV, idxList);
       ReplaceUsesForLoweredUDT(GEP, NewGEP);
+      dxilutil::MergeGepUse(NewGEP);
```

Comment:
Do we need to call `MergeGepUse` for the `GEPOperator` below as well? Seems like it would have the same problem (and points to a mising test case?).

Reply:
That case handles constant GEP.  Wouldn't constant GEP already be in a merged state?

Reply:
The purpose here was to avoid DXIL that had a GEP instruction returning an intermediate type which was illegal on its own in DXIL (DxilValidation).  Multi-dim array is legal within a struct, but not outside the struct, so if multiple GEPs are used to drill into a struct, an intermediate type could appear illegal to DxilValidation.  These have to be GEP instructions, otherwise the GEP is a constant expression passed to a load or store instruction, and DxilValidation only cares about the type loaded or stored, which should be scalar from SROA.

Reply:
Constant GEPs can have multiple levels of gep as well, its not necessarily a single level gep.

Reply:
Ok, so I would be able to add the MergeGepUse to the constant case as well, but I don't think it would cause any failure to have a multi-level constant GEP, and I'm not confident I could create a test case for it.  For this node record case, it'll never be a constant.  It might require a carefully crafted AS or MS test case to trigger it, using a groupshared mesh shader payload.

In any case, I don't think multi-level constant GEP will cause any problem for DxilValidation, so the test would only be making sure the output IR looks a particular way.  I don't think it would be fixing an actual bug.

Should I try to do it anyway?